File: src/node/tui/compositor.ts
// Compositor: manages independent panel buffers with dirty-flag rendering
// Each panel has its own update cycle; the compositor tick only redraws dirty regions.
// Exception: RainLayer writes directly to stdout via its own setInterval,
// bypassing the compositor tick — this decouples rain from the rendering pipeline
// so it never freezes during API fetches.

import type { PanelSession } from "./panel.js";
import { buildStatsGrid } from "./panel.js";
import { RainState } from "./rain.js";
import { dim, stripAnsi } from "./colors.js";

export interface PanelBuffer {
  render(): string[];
  readonly dirty: boolean;
}

export interface CompositorOptions {
  noRain?: boolean;
  getTermWidth: () => number;
  getTermRows: () => number;
}

const RAIN_TICK_MS = 107;
const COUNTDOWN_TICK_MS = 1000;
export const COMPACT_THRESHOLD = 60;

// Right-margin rain: minimum usable width, and gutter columns kept between the
// widest content line and the first rain column so glyphs don't touch the table.
const MIN_RAIN_COLS = 10;
const RAIN_GUTTER = 2;

// --- TablePanel ---
class TablePanel implements PanelBuffer {
  dirty = false;

  private lines: string[] = [];

  update(lines: string[]): void {
    this.lines = lines;
    this.dirty = true;
  }

  render(): string[] {
    this.dirty = false;
    return this.lines;
  }
}

// --- StatsPanel ---
class StatsPanel implements PanelBuffer {
  dirty = false;

  private lines: string[] = [];

  update(session: PanelSession, todayCost: number): void {
    this.lines = buildStatsGrid(session, todayCost);
    this.dirty = true;
  }

  render(): string[] {
    this.dirty = false;
    return this.lines;
  }

  getLines(): string[] {
    return this.lines;
  }
}

// --- StatusPanel ---
class StatusPanel implements PanelBuffer {
  dirty = false;

  private content = "";
  private countdown: number | null = null;
  private termWidth = 80;

  updateCountdown(value: number | null, termWidth: number): void {
    this.countdown = value;
    this.termWidth = termWidth;
    this.dirty = true;
  }

  render(): string[] {
    this.dirty = false;
    const parts: string[] = [];

    if (this.countdown === null) {
      parts.push(dim("Refreshing..."));
    } else {
      parts.push(dim(`Next refresh: ${this.countdown}s`));
    }

    parts.push(dim("\u21B5 refresh \u00B7 q quit"));

    // Progressive truncation
    let line = parts.join(dim(" \u00B7 "));
    while (stripAnsi(line).length > this.termWidth && parts.length > 1) {
      parts.pop();
      line = parts.join(dim(" \u00B7 "));
    }

    this.content = line;
    return [this.content];
  }

  getContent(): string {
    return this.content;
  }
}

// --- RainLayer ---
// RainLayer uses cursor-positioned writes as an overlay and does NOT trigger
// a full recomposite. Its tick/render cycle runs on its own setInterval,
// independent of the compositor tick.
class RainLayer {
  private rain: RainState | null = null;
  private startRow = 0;
  private enabled = false;

  setup(cols: number, availableRows: number, startRow: number, startCol = 0): void {
    if (availableRows <= 0 || cols <= 0) {
      this.enabled = false;
      this.rain = null;
      return;
    }
    this.enabled = true;
    this.startRow = startRow;
    if (!this.rain) {
      this.rain = new RainState(cols, availableRows, startCol);
    } else {
      this.rain.resize(cols, availableRows, startCol);
    }
  }

  disable(): void {
    this.enabled = false;
    this.rain = null;
  }

  tick(): void {
    if (!this.enabled || !this.rain) return;
    this.rain.tick();
  }

  renderDirect(): string {
    if (!this.enabled || !this.rain) return "";
    return this.rain.render(this.startRow);
  }

  isEnabled(): boolean {
    return this.enabled;
  }
}

// --- Compositor ---
export class Compositor {
  readonly table: TablePanel;
  readonly stats: StatsPanel;
  readonly status: StatusPanel;
  readonly rainLayer: RainLayer;

  private opts: CompositorOptions;
  private rainTimer: ReturnType<typeof setInterval> | null = null;
  private countdownTimer: ReturnType<typeof setTimeout> | null = null;
  private countdownValue = 0;

  // Cached data for resize re-rendering
  private lastTableLines: string[] = [];
  private lastSession: PanelSession | null = null;
  private lastTodayCost = 0;

  constructor(opts: CompositorOptions) {
    this.opts = opts;
    this.table = new TablePanel();
    this.stats = new StatsPanel();
    this.status = new StatusPanel();
    this.rainLayer = new RainLayer();
  }

  start(): void {
    // Rain tick: independent ~107ms animation (~75% as fast as the original 80ms rate)
    if (!this.opts.noRain) {
      this.rainTimer = setInterval(() => {
        this.rainLayer.tick();
        const rainOutput = this.rainLayer.renderDirect();
        if (rainOutput) process.stdout.write(rainOutput);
      }, RAIN_TICK_MS);
    }
  }

  stop(): void {
    if (this.rainTimer !== null) {
      clearInterval(this.rainTimer);
      this.rainTimer = null;
    }
    if (this.countdownTimer !== null) {
      clearTimeout(this.countdownTimer);
      this.countdownTimer = null;
    }
  }

  startCountdown(seconds: number, onExpire: () => void): void {
    this.countdownValue = seconds;
    this.status.updateCountdown(seconds, this.opts.getTermWidth());
    this.renderStatus();

    const tickDown = () => {
      this.countdownValue--;
      if (this.countdownValue <= 0) {
        onExpire();
      } else {
        this.status.updateCountdown(this.countdownValue, this.opts.getTermWidth());
        this.renderStatus();
        this.countdownTimer = setTimeout(tickDown, COUNTDOWN_TICK_MS);
      }
    };

    this.countdownTimer = setTimeout(tickDown, COUNTDOWN_TICK_MS);
  }

  cancelCountdown(): void {
    if (this.countdownTimer !== null) {
      clearTimeout(this.countdownTimer);
      this.countdownTimer = null;
    }
  }

  setRefreshing(): void {
    this.status.updateCountdown(null, this.opts.getTermWidth());
    this.renderStatus();
  }

  /** Called after API poll with new table data and panel session */
  updateAfterPoll(
    tableLines: string[],
    session: PanelSession,
    todayCost: number,
  ): void {
    // Cache data for resize re-rendering
    this.lastTableLines = tableLines;
    this.lastSession = session;
    this.lastTodayCost = todayCost;

    this.layoutAndUpdate(tableLines, session, todayCost);
  }

  /** Re-layout and flush using cached data (called on terminal resize) */
  rerender(): void {
    if (this.lastTableLines.length === 0) return;
    this.layoutAndUpdate(
      this.lastTableLines,
      this.lastSession!,
      this.lastTodayCost,
    );
    this.flush();
  }

  private layoutAndUpdate(
    tableLines: string[],
    session: PanelSession,
    todayCost: number,
  ): void {
    const tw = this.opts.getTermWidth();
    const compact = tw < COMPACT_THRESHOLD;

    this.table.update(tableLines);

    if (!compact) {
      this.stats.update(session, todayCost);
    }

    // Compute content height for rain zone calculation
    const statsLines = compact ? [] : this.stats.getLines();
    const contentHeight = statsLines.length + tableLines.length;
    const maxContentWidth = this.computeMaxContentWidth([...statsLines, ...tableLines]);

    this.setupRainZone(contentHeight, maxContentWidth);
  }

  /**
   * Lay out the rain zone against loading-skeleton geometry so rain animates
   * before the first poll completes. Derives content height and max content
   * width from the skeleton lines and reuses the shared rain-zone setup.
   */
  layoutForSkeleton(lines: string[]): void {
    const contentHeight = lines.length;
    const maxContentWidth = this.computeMaxContentWidth(lines);
    this.setupRainZone(contentHeight, maxContentWidth);
  }

  /** Shared rain-zone computation used by both poll layout and skeleton layout. */
  private setupRainZone(contentHeight: number, maxContentWidth: number): void {
    const tw = this.opts.getTermWidth();
    const compact = tw < COMPACT_THRESHOLD;

    const footerRow = 1;
    const termRows = this.opts.getTermRows();
    const availableRainRows = termRows - contentHeight - footerRow;
    const wantRain = !this.opts.noRain && !compact;

    if (wantRain && availableRainRows > 0) {
      // Below-content rain (full width, rows below content)
      this.rainLayer.setup(tw, availableRainRows, contentHeight + 1);
    } else if (wantRain) {
      // Right-margin rain: render in columns past the content width, keeping a
      // gutter so glyphs don't visually collide with the table edge.
      const marginCols = tw - maxContentWidth - RAIN_GUTTER;
      if (marginCols >= MIN_RAIN_COLS) {
        const rainRows = termRows - footerRow;
        this.rainLayer.setup(marginCols, rainRows, 1, maxContentWidth + RAIN_GUTTER);
      } else {
        this.rainLayer.disable();
      }
    } else {
      this.rainLayer.disable();
    }
  }

  private computeMaxContentWidth(lines: string[]): number {
    let maxWidth = 0;
    for (const line of lines) {
      const len = stripAnsi(line).length;
      if (len > maxWidth) maxWidth = len;
    }
    return maxWidth;
  }

  /** Force a full composite and flush to screen */
  flush(): void {
    const tw = this.opts.getTermWidth();
    const compact = tw < COMPACT_THRESHOLD;

    const statsLines = compact ? [] : this.stats.render();
    const tableLines = this.table.render();
    const allLines = [...statsLines, ...tableLines];

    // Cursor home, write output
    process.stdout.write("\x1b[H");
    for (const line of allLines) {
      process.stdout.write(line + "\x1b[K\n");
    }
    // Clear stale trailing lines
    process.stdout.write("\x1b[J");

    // Write footer
    const statusLines = this.status.render();
    if (statusLines.length > 0) {
      writeFooterLine(statusLines[0], this.opts.getTermRows());
    }

    // Restore the rain zone: flush's \x1b[J/\x1b[K clears erased every drawn
    // rain cell. RainState.render() rewrites every occupied cell each frame, so
    // re-emitting it here fully restores the rain without waiting for the next
    // rain tick (which would otherwise leave a per-poll blink).
    const rainOut = this.rainLayer.renderDirect();
    if (rainOut) process.stdout.write(rainOut);
  }

  /**
   * Render the footer status line immediately. Push-driven — called after every
   * countdown/refresh state change (replaces the former periodic 16ms tick).
   */
  private renderStatus(): void {
    const statusLines = this.status.render();
    if (statusLines.length > 0) {
      writeFooterLine(statusLines[0], this.opts.getTermRows());
    }
  }
}

// Shared utilities
function writeFooterLine(content: string, rows: number): void {
  process.stdout.write(`\x1b[${rows};1H\x1b[K${content}`);
}

File: src/go/internal/watch/compositor.go
package watch

import (
	"strings"
	"time"
	"unicode/utf8"

	"github.com/sahil87/tu/internal/render/ansi"
)

// The compositor constants (the TS compositor.ts / watch.ts).
const (
	// CompactThreshold: below this width the stats grid and rain drop out and
	// the compact table renders (the TS COMPACT_THRESHOLD).
	CompactThreshold = 60
	// RainTick is the rain animation's interval (the TS RAIN_TICK_MS — 75% of
	// the original 80 ms rate).
	RainTick = 107 * time.Millisecond
	// CountdownTick is the countdown timer's 1 s step (the TS
	// COUNTDOWN_TICK_MS).
	CountdownTick = time.Second
	// minRainCols is the right-margin zone's minimum usable width (the TS
	// MIN_RAIN_COLS).
	minRainCols = 10
	// rainGutter is the gap kept between the widest content line and the
	// first rain column in right-margin mode (the TS RAIN_GUTTER).
	rainGutter = 2
	// footerRows is the footer band's height.
	footerRows = 1
	// MaxRows is the watch row budget (the TS maxRows: 15 — a constant, not
	// "rows that fit the terminal height"; the spec sentence is a G0 flag).
	MaxRows = 15
)

// RainZone is the rain layer's geometry: Enabled, the zone's size, the
// 1-based StartRow and the 0-based StartCol (the right-margin offset).
type RainZone struct {
	Enabled  bool
	Cols     int
	Rows     int
	StartRow int
	StartCol int
}

// Layout is one frame's layout record: the compact switch, the content lines
// (stats nil when compact), the rain zone, and the terminal height the footer
// positions against.
type Layout struct {
	Compact bool
	Stats   []string
	Table   []string
	Rain    RainZone
	Rows    int // terminal rows (footer position)
}

// Lay reproduces the TS layoutAndUpdate + setupRainZone: compact when cols <
// 60 (stats dropped); contentHeight = len(stats) + len(table); available =
// rows − contentHeight − 1; wantRain = !noRain && !compact. Below-content
// zone {cols, available, startRow contentHeight+1, startCol 0} when wantRain
// && available > 0; else the right-margin zone {cols − maxContentWidth − 2,
// rows − 1, startRow 1, startCol maxContentWidth + 2} when that width is ≥
// 10; else disabled. maxContentWidth is the max StripANSI rune width over the
// content lines.
func Lay(statsLines, tableLines []string, cols, rows int, noRain bool) Layout {
	compact := cols < CompactThreshold
	l := Layout{Compact: compact, Rows: rows, Table: tableLines}
	if !compact {
		l.Stats = statsLines
	}
	contentHeight := len(l.Stats) + len(l.Table)
	maxContentWidth := 0
	for _, line := range l.Stats {
		if n := utf8.RuneCountInString(ansi.StripANSI(line)); n > maxContentWidth {
			maxContentWidth = n
		}
	}
	for _, line := range l.Table {
		if n := utf8.RuneCountInString(ansi.StripANSI(line)); n > maxContentWidth {
			maxContentWidth = n
		}
	}

	available := rows - contentHeight - footerRows
	wantRain := !noRain && !compact
	switch {
	case wantRain && available > 0:
		l.Rain = RainZone{Enabled: true, Cols: cols, Rows: available, StartRow: contentHeight + 1}
	case wantRain:
		marginCols := cols - maxContentWidth - rainGutter
		if marginCols >= minRainCols {
			l.Rain = RainZone{Enabled: true, Cols: marginCols, Rows: rows - footerRows, StartRow: 1, StartCol: maxContentWidth + rainGutter}
		}
	}
	return l
}

// LaySkeleton is the TS layoutForSkeleton: the skeleton's own lines are the
// whole content (no stats/table split).
func LaySkeleton(lines []string, cols, rows int, noRain bool) Layout {
	return Lay(nil, lines, cols, rows, noRain)
}

// Frame reproduces the TS flush: cursor home; each content line (stats then
// table) + `\x1b[K\n`; `\x1b[J`; the footer at the last row
// (`\x1b[{rows};1H\x1b[K{footer}`); then the current rain frame re-emitted —
// the flush's clears erased every drawn rain cell, and the rain renderer
// rewrites every occupied cell, so the re-emit restores them in the same
// write and polls do not blink the rain.
func (l Layout) Frame(footer string, rain string) []byte {
	var b strings.Builder
	b.WriteString("\x1b[H")
	for _, line := range l.Stats {
		b.WriteString(line + "\x1b[K\n")
	}
	for _, line := range l.Table {
		b.WriteString(line + "\x1b[K\n")
	}
	b.WriteString("\x1b[J")
	b.WriteString(FooterLine(footer, l.Rows))
	b.WriteString(rain)
	return []byte(b.String())
}

// FooterLine is the footer-only write (the TS writeFooterLine — renderStatus
// on every countdown change, push-driven, never on a periodic tick).
func FooterLine(footer string, rows int) string {
	return "\x1b[" + itoa(rows) + ";1H\x1b[K" + footer
}
